File: bound/kt/src/main/kotlin/tbdex/sdk/httpclient/Offerings.kt
package tbdex.sdk.httpclient

import tbdex.sdk.TbdexException
import tbdex.sdk.rust.getOfferings as rustCoreGetOfferings
import tbdex.sdk.resources.Offering

fun getOfferings(pfiDidUri: String): List<Offering> {
    try {
        val rustCoreOfferings = rustCoreGetOfferings(pfiDidUri)
        return rustCoreOfferings.map { Offering.fromRustCoreOffering(it) }
    } catch (e: tbdex.sdk.rust.TbdexException.Exception) {
        throw TbdexException.fromRustCore(e)
    }
}

File: bindings/tbdex_uniffi/src/http_client/offerings.rs
use crate::{errors::Result, get_rt, resources::offering::Offering};
use std::sync::{Arc, RwLock};

pub fn get_offerings(pfi_did_uri: String) -> Result<Vec<Arc<Offering>>> {
    let rt = get_rt()?;
    let inner_offerings =
        rt.block_on(tbdex::http_client::offerings::get_offerings(&pfi_did_uri))?;

    let offerings = inner_offerings
        .into_iter()
        .map(|o| Arc::new(Offering(Arc::new(RwLock::new(o)))))
        .collect();

    Ok(offerings)
}
